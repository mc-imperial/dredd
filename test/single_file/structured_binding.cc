void foo() {
  int a[2] = {1, 2};
  if (true) {
    auto [x, y] = a;
    auto& [xr, yr] = a;
    xr = y++;
    yr = x++;
  }
}
