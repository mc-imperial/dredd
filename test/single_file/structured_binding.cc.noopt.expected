#include <cinttypes>
#include <cstddef>
#include <functional>
#include <string>


#ifdef _MSC_VER
#define thread_local __declspec(thread)
#elif __APPLE__
#define thread_local __thread
#endif

static thread_local bool __dredd_some_mutation_enabled = true;
static bool __dredd_enabled_mutation(int local_mutation_id) {
  static thread_local bool initialized = false;
  static thread_local uint64_t enabled_bitset[1];
  if (!initialized) {
    bool some_mutation_enabled = false;
    const char* dredd_environment_variable = std::getenv("DREDD_ENABLED_MUTATION");
    if (dredd_environment_variable != nullptr) {
      std::string contents(dredd_environment_variable);
      while (true) {
        size_t pos = contents.find(",");
        std::string token = (pos == std::string::npos ? contents : contents.substr(0, pos));
        if (!token.empty()) {
          int value = std::stoi(token);
          int local_value = value - 0;
          if (local_value >= 0 && local_value < 60) {
            enabled_bitset[local_value / 64] |= (1 << (local_value % 64));
            some_mutation_enabled = true;
          }
        }
        if (pos == std::string::npos) {
          break;
        }
        contents.erase(0, pos + 1);
      }
    }
    initialized = true;
    __dredd_some_mutation_enabled = some_mutation_enabled;
  }
  return (enabled_bitset[local_mutation_id / 64] & (1 << (local_mutation_id % 64))) != 0;
}

static int& __dredd_replace_binary_operator_Assign_arg1_int_arg2_int(int& arg1, std::function<int()> arg2, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg1 = arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg1 += arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return arg1 &= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return arg1 /= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return arg1 *= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return arg1 |= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return arg1 %= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 6)) return arg1 <<= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 7)) return arg1 >>= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 8)) return arg1 -= arg2();
  if (__dredd_enabled_mutation(local_mutation_id + 9)) return arg1 ^= arg2();
  return arg1 = arg2();
}

static int __dredd_replace_unary_operator_PostInc_int(std::function<int&()> arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg()++;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return arg()--;
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~arg();
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -arg();
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return !arg();
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return arg();
  return arg()++;
}

static int __dredd_replace_expr_int(std::function<int()> arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg();
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -(arg());
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 0;
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return 1;
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -1;
  return arg();
}

static int __dredd_replace_expr_int(int arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return ~(arg);
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return -(arg);
  if (__dredd_enabled_mutation(local_mutation_id + 3)) return 0;
  if (__dredd_enabled_mutation(local_mutation_id + 4)) return 1;
  if (__dredd_enabled_mutation(local_mutation_id + 5)) return -1;
  return arg;
}

static bool __dredd_replace_expr_bool(bool arg, int local_mutation_id) {
  if (!__dredd_some_mutation_enabled) return arg;
  if (__dredd_enabled_mutation(local_mutation_id + 0)) return !(arg);
  if (__dredd_enabled_mutation(local_mutation_id + 1)) return true;
  if (__dredd_enabled_mutation(local_mutation_id + 2)) return false;
  return arg;
}

void foo() {
  int a[2] = {__dredd_replace_expr_int(1, 0), __dredd_replace_expr_int(2, 6)};
  if (!__dredd_enabled_mutation(59)) { if (__dredd_replace_expr_bool(true, 12)) {
    auto [x, y] = a;
    auto& [xr, yr] = a;
    if (!__dredd_enabled_mutation(36)) { __dredd_replace_binary_operator_Assign_arg1_int_arg2_int(xr , [&]() -> int { return static_cast<int>(__dredd_replace_expr_int([&]() -> int { return static_cast<int>(__dredd_replace_unary_operator_PostInc_int([&]() -> int& { return static_cast<int&>(y); }, 15)); }, 20)); }, 26); }
    if (!__dredd_enabled_mutation(58)) { __dredd_replace_binary_operator_Assign_arg1_int_arg2_int(yr , [&]() -> int { return static_cast<int>(__dredd_replace_expr_int([&]() -> int { return static_cast<int>(__dredd_replace_unary_operator_PostInc_int([&]() -> int& { return static_cast<int&>(x); }, 37)); }, 42)); }, 48); }
  } }
}
